### PR TITLE
Fix Concurrency Bug In CommitToParent

### PR DESCRIPTION
**Pull request by dboehm-avalabs**
CommitToParent can happen concurrently so it may cause the parent of a view to change during its own attempt to commit to its parent(see the new unit tests).  This isn't an issue for CommitToDB as it uses the commit lock.

### CODE REVIEW

File: x/merkledb/trieview.go [hunk 1]
```diff
@@ -222,10 +222,8 @@ func (t *trieView) calculateNodeIDs(ctx context.Context) error {
 
 	// ensure that the view under this one is up-to-date before potentially pulling in nodes from it
 	// getting the Merkle root forces any unupdated nodes to recalculate their ids
-	if t.parentTrie != nil {
```
Comment (dboehm-avalabs) [inline x/merkledb/trieview.go:225]: detects as race and no longer needed.  Hold over from when views could have no parent view.

File: x/merkledb/db.go [hunk 2]
```diff
@@ -752,8 +752,8 @@ func (db *Database) commitChanges(ctx context.Context, trieToCommit *trieView) e
 	// invalidate all child views except for the view being committed
 	db.invalidateChildrenExcept(trieToCommit)
 
-	// move any child views of the committed trie onto the db
-	db.moveChildViewsToDB(trieToCommit)
+	// move any child views of the committed trie onto the db after changes have been made
+	defer db.moveChildViewsToDB(trieToCommit)
```
Comment (danlaine) [inline x/merkledb/db.go:756]: Why is this in a defer now? We hold the database lock here so does it make a difference whether this is deferred?

Comment (dboehm-avalabs) [reply x/merkledb/db.go:756]: Yes, it was updating the children's parent before the data was in the cache, so children could get stale data.

Comment (dboehm-avalabs) [reply x/merkledb/db.go]: turns out this worked but wasn't the underlying issue.

File: x/merkledb/db.go [hunk 3]
```diff
@@ -752,7 +752,7 @@ func (db *Database) commitChanges(ctx context.Context, trieToCommit *trieView) e
 	// invalidate all child views except for the view being committed
 	db.invalidateChildrenExcept(trieToCommit)
 
-	// move any child views of the committed trie onto the db
+	// move any child views of the committed trie onto the db after changes have been made
```
Comment (danlaine) [inline x/merkledb/db.go:755]:
```suggestion
	// move any child views of the committed trie onto the db
```

File: x/merkledb/trieview.go [hunk 4]
```diff
@@ -567,11 +565,8 @@ func (t *trieView) commitChanges(ctx context.Context, trieToCommit *trieView) er
 
 // CommitToParent commits the changes from this view to its parent Trie
 func (t *trieView) CommitToParent(ctx context.Context) error {
-	// if we are about to write to the db, then we to hold the commitLock
-	if t.getParentTrie() == t.db {
-		t.db.commitLock.Lock()
-		defer t.db.commitLock.Unlock()
-	}
+	t.db.commitLock.Lock()
```
Comment (dboehm-avalabs) [inline x/merkledb/trieview.go:568]: Prevents multiple commits from occurring at the same time until I can figure out a way to manage the safety of this.

**Review by danlaine — APPROVED** [review]

**Review by darioush — APPROVED** [review]

**Pull request merged**